go/types/graph_builder.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/dustin/go-humanize"
+)
+
+type GraphBuilder struct {
+	oc      opCache
+	vrw     ValueReadWriter
+	stack   graphStack
+	verbose bool
+}
+
+// NewGraphBuilder() returns an object that is used in building a graph of
+// objects that are stored on disk and then assembled into a noms collection.
+// Graphs can exist of nested maps, with maps, sets and lists at the outermost
+// levels of the graph. Any of the collections (nested or leaf) can contain onto
+// any type of Value. |rootKind| indicates what the topLevel collection should be.
+// It must be one of: MapKind, SetKind, or ListKind. If this GraphBuilder is
+// intended to build nested maps, |rootKind| must be MapKind.
+func NewGraphBuilder(vrw ValueReadWriter, rootKind NomsKind, verbose bool) *GraphBuilder {
```

Comment:
Is this abstraction intended to be thread safe (usable from multiple goroutines)?